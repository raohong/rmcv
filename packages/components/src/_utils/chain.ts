const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
