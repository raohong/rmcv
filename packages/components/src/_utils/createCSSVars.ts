import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
