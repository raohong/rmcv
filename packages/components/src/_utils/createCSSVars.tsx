import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
