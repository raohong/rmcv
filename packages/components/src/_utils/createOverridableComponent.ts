import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
