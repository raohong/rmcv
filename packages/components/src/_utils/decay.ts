import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
