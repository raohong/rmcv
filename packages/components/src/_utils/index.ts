export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
