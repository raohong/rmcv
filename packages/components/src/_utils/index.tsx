export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
