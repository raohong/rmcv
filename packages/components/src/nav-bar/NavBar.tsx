import classNames from 'classnames';
import { useMeasure, useMergeRefs } from '@rmc-vant/hooks';
import { isEmpty } from '@rmc-vant/utils';
import { ArrowLeft } from '@rmc-vant/icons';
import React, { useRef } from 'react';
import { useConfigContext } from '../config-provider';
import SafeArea from '../safe-area';
import type { NavBarProps } from './interface';
import Touchable from '../touchable';

const NavBar = React.forwardRef<HTMLDivElement, NavBarProps>(
  (
    {
      title,
      leftText,
      rightText,
      left,
      right,
      leftArrow,
      onClickLeft,
      onClickRight,
      fixed,
      placeholder,
      zIndex,
      className,
      style,
      safeAreaInsetTop = true,
      border = true,
      children,
      ...rest
    },
    ref,
  ) => {
    const { getPrefixCls } = useConfigContext();
    const domRef = useRef<HTMLDivElement | null>(null);
    const mergedRef = useMergeRefs(domRef, ref);
    const {
      data: { height },
    } = useMeasure({ ref: domRef });

    const baseCls = getPrefixCls('nav-bar');

    const renderLeft = () => {
      if (isEmpty(left)) {
        return (
          <>
            {!!leftArrow && <ArrowLeft className={`${baseCls}-arrow-icon`} />}
            <span className={`${baseCls}-text`}>{leftText}</span>
          </>
        );
      }

      return left;
    };

    const renderRight = () => {
      let target: React.ReactNode;
      if (isEmpty(right)) {
        target = isEmpty(rightText) ? null : (
          <span className={`${baseCls}-text`}>{rightText}</span>
        );
      } else {
        target = right;
      }

      return isEmpty(target) ? null : (
        <Touchable
          activeClassName={`${baseCls}-haptics-feedback`}
          onClick={onClickRight}
          className={`${baseCls}-right`}
        >
          {target}
        </Touchable>
      );
    };

    return (
      <>
        {!!placeholder && !!fixed && (
          <SafeArea
            className={`${baseCls}-placeholder`}
            disabled={!safeAreaInsetTop}
            style={{ height: height === 0 ? undefined : height }}
          />
        )}
        <SafeArea
          className={classNames(
            baseCls,
            {
              [`${baseCls}-fixed`]: fixed,
              [`${baseCls}-border`]: border,
            },
            className,
          )}
          disabled={!safeAreaInsetTop}
          ref={mergedRef}
          style={{
            zIndex,
            ...style,
          }}
          {...rest}
        >
          <Touchable
            activeClassName={`${baseCls}-haptics-feedback`}
            onClick={onClickLeft}
            className={`${baseCls}-left`}
          >
            {renderLeft()}
          </Touchable>
          <div className={`${baseCls}-title`}>{title}</div>
          {renderRight()}
        </SafeArea>
      </>
    );
  },
);

export default NavBar;
