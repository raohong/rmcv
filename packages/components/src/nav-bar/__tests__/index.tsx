import React from 'react';
import { render, screen } from '@testing-library/react';
import { getPrefixCls } from '../../_utils';
import NavBar from '../NavBar';

const testId = 'nav-bar';

test('render correctly', () => {
  const tree = render(<NavBar />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with title', () => {
  render(<NavBar title="title" />);

  expect(screen.getByText('title')).toBeInTheDocument();
});

test('render with leftArrow', () => {
  render(<NavBar title="title" data-testid={testId} leftArrow />);

  expect(screen.getByRole('img')).toHaveClass(getPrefixCls('nav-bar-arrow-icon'));
});

test('render with leftText', () => {
  render(<NavBar leftText="left-text" />);

  expect(screen.getByText('left-text')).toBeInTheDocument();
});

test('render with rightText', () => {
  render(<NavBar rightText="right-text" />);

  expect(screen.getByText('right-text')).toBeInTheDocument();
});

test('render with left', () => {
  render(<NavBar left={<span data-testid="left" />} />);

  expect(screen.getByTestId('left')).toBeInTheDocument();
});

test('render with right', () => {
  render(<NavBar right={<span data-testid="right" />} />);

  expect(screen.getByTestId('right')).toBeInTheDocument();
});

test('render with zIndex', () => {
  render(<NavBar zIndex={2} data-testid={testId} />);

  expect(screen.getByTestId(testId)).toHaveStyle({
    'z-index': 2,
  });
});

test('render with placeholder', () => {
  const com = render(<NavBar placeholder />);

  expect(
    com.container.querySelector(`.${getPrefixCls('placeholder')}`),
  ).not.toBeInTheDocument();

  com.rerender(<NavBar placeholder fixed />);

  expect(
    com.container.querySelector(`.${getPrefixCls('placeholder')}`),
  ).not.toBeInTheDocument();
});
