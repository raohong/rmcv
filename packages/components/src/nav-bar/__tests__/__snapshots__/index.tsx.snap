// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    class="rmcv-nav-bar rmcv-nav-bar-border rmcv-safe-area-bottom"
  >
    <div
      class="rmcv-nav-bar-left rmcv-touchable"
      style="cursor: cursor;"
    >
      <span
        class="rmcv-nav-bar-text"
      />
    </div>
    <div
      class="rmcv-nav-bar-title"
    />
  </div>
</DocumentFragment>
`;
